core/client/fs/src/main/java/alluxio/client/file/dora/netty/NettyClientStateMachine.java

```diff
@@ -0,0 +1,705 @@
+      mStateMachine.configuration().generateDotFileInto(outFile, /* printLabels */ true);
+    }
+  }
+
+  /**
+   * Gets the exception during execution.
+   *
+   * @return exception, or null if there is none
+   */
+  @Nullable
+  public PartialReadException getException() {
+    if (mLastExceptionTrigger == null) {
+      return null;
+    }
+    switch (mLastExceptionTrigger) {
+      case INTERRUPTED:
+        return new PartialReadException(mLength, mBytesRead, CauseType.INTERRUPT, mLastException);
+      case OUTPUT_ERROR:
+        return new PartialReadException(mLength, mBytesRead, CauseType.OUTPUT, mLastException);
+      case TIMEOUT:
+        return new PartialReadException(mLength, mBytesRead, CauseType.TIMEOUT, mLastException);
+      case SERVER_ERROR:
+        return new PartialReadException(
+            mLength, mBytesRead, CauseType.SERVER_ERROR, mLastException);
+      case CHANNEL_ERROR:
+        // fallthrough
+      case CHANNEL_UNAVAILABLE:
+        return new PartialReadException(
+            mLength, mBytesRead, CauseType.TRANSPORT_ERROR, mLastException);
+      default:
+        throw new IllegalStateException("unexpected trigger type: " + mLastExceptionTrigger);
+    }
+  }
+
+  /**
+   * Helper method to allow firing triggers within state handler methods.
+   * If the triggers are fired directly within state handler methods, they will likely make
+   * recursive calls and blow up the stack.
+   * @param trigger the next trigger to fire
+   */
+  void fireNext(Trigger trigger) {
+    mNextTrigger.set(() -> mStateMachine.fire(trigger));
+  }
+
+  <Arg0T> void fireNext(TriggerWithParameters1<Arg0T, Trigger> trigger, Arg0T arg0) {
+    mNextTrigger.set(() -> mStateMachine.fire(trigger, arg0));
+  }
+
+  void acquireNettyChannel() {
+    try {
+      mChannel = mContext.acquireNettyChannel(mAddress);
+      mChannel.pipeline().addLast(new PacketReadHandler(mQueue, mMaxPacketsInFlight));
+    } catch (IOException ioe) {
+      fireNext(mTriggers.mChannelUnavailable, ioe);
+      return;
+    }
+    fireNext(Trigger.CHANNEL_AVAILABLE);
+  }
+
+  void sendRequest() {
+    Preconditions.checkNotNull(mChannel, "channel has not been acquired");
+    Protocol.ReadRequest readRequest = mRequestBuilder.get().clearCancel().build();
+    mChannel.writeAndFlush(new RPCProtoMessage(new ProtoMessage(readRequest)))
+        .addListener((ChannelFutureListener) future -> {
+          if (!future.isSuccess()) {
+            // Note: cannot call fireNext(Trigger.CHANNEL_ERROR, future.cause()) directly
+            // as the callback is called on a Netty I/O thread, it would bypass the blocking queue
+            // and create a race condition with the thread the state machine is executing on
+            mQueue.offer(Payload.channelError(future.cause()));
+          }
+        });
+  }
+
+  void pollResponseFromQueue() {
+    if (!tooManyPacketsPending(mQueue, mMaxPacketsInFlight)) {
+      NettyUtils.enableAutoRead(mChannel);
```

Comment:
todo: `mChannel` can be `null` in some states. Consider re-organize the code so that each state is represented by a dedicated class, and in states where a channel isn't available, state handler methods cannot obtain a null reference at all.